### ISSUE TITLE
iree-compile crash / codgen error involving vector.gather op

### ISSUE THREAD

**Issue by astrotuna201** [iree-org/iree#11126]
### What happened?

Following on from the resolved #11062 (thank you for fixing!), there is another compiler crash:

```
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/numerics.py:37:0: error: 'vector.gather' op operand #2 must be vector of integer or index values, but got 'index'
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/operators.py:146:0: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/operators.py:146:0: error: failed to serialize executables
compilation failed
```

```
Invoked with:
 iree-compile /Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false
```

### Steps to reproduce your issue

```core-command-line.txt```:
```
/Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/wiggles/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false --mlir-pass-pipeline-crash-reproducer=savedir/core-reproducer.mlir
```

reproducer from ```core-reproducer.mlir```:

[core-reproducer.mlir.txt](https://github.com/iree-org/iree/files/9988026/core-reproducer.mlir.txt)

### What component(s) does this issue relate to?

Compiler

### Version information

[iree candidate candidate-20221110.323](https://github.com/iree-org/iree/releases/tag/candidate-20221110.323)

**Comment by rsuderman** [comment]
Fix sent for review upstream https://reviews.llvm.org/D137990

**Comment by astrotuna201** [comment]
@rsuderman 

Many thanks for trying to fix! However, even with [candidate-20221116.329](https://github.com/iree-org/iree/releases/tag/candidate-20221116.329) the compilation still fails. Did you manage to compiler the reproducer ok? 

I now get ```error: 'iree_linalg_ext.scatter' op mismatch in shape of indices and update value at dim#0```

A simpler snippet to demonstrate the same failure is happening on mhlo.scatter 
with

```
#loc = loc(unknown)
module @jit_calc_grid_spacings_kernel {
  func.func public @main(%arg0: tensor<94xf32> loc(unknown)) -> (tensor<94xf32>) {
    %6 = mhlo.constant dense<2> : tensor<i32> loc(#loc)
    %7 = "mhlo.dynamic_slice"(%arg0, %6) {slice_sizes = dense<2> : tensor<1xi64>} : (tensor<94xf32>, tensor<i32>) -> tensor<2xf32> loc(#loc)
    %8 = mhlo.constant dense<92> : tensor<i32> loc(#loc)
    %9 = "mhlo.broadcast_in_dim"(%8) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<1xi32> loc(#loc)
    %10 = "mhlo.scatter"(%arg0, %9, %7) ({
    ^bb0(%arg12: tensor<f32> loc(unknown), %arg13: tensor<f32> loc(unknown)):
      mhlo.return %arg13 : tensor<f32> loc(#loc)
    }) {indices_are_sorted = true, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0], scatter_dims_to_operand_dims = [0]>, unique_indices = true} : (tensor<94xf32>, tensor<1xi32>, tensor<2xf32>) -> tensor<94xf32> loc(#loc)
    return %10: tensor<94xf32>
  }
}
```

which is an abbreviated first section of the initial reproducer.
For context: the code is generated from the iree backend of jax. However, when using the numpy backend with xla it runs fine ...
